### PR TITLE
Fixes #9105: Add conditions on features in ncf code loading

### PR DESCRIPTION
**Pull request by peckpeck**
https://www.rudder-project.org/redmine/issues/9105

### CODE REVIEW

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
Commit modified

**Comment by peckpeck** [comment]
Commit modified

File: tree/10_ncf_internals/list-compatible-inputs [hunk 1]
```diff
@@ -5,25 +5,56 @@ set -e
...
+else
+  cfengine_version="$1"
+  framework_path="$2"
+  # see comment below on the last parameters
+  shift 2
+fi
+
 # split version numbers
 version_regex='\([0-9][0-9]*\)\.\([0-9][0-9]*\).*'
 cfengine_major=`printf "${cfengine_version}\n" | sed -e "s/${version_regex}/\\1/"`
```
Comment (amousset) [inline tree/10_ncf_internals/list-compatible-inputs]: Shouldn't we check cfengine_version existence?

File: tree/10_ncf_internals/list-compatible-inputs [hunk 2]
```diff
@@ -62,31 +93,68 @@ do
...
+            include_reqs=1 # false in shell
+          fi
+        done
+      else
+        # no capabilities required = accept
+        include_reqs=0 # true for a shell
+      fi
+
+      # merge and exclude if necessary
+      if [ ${include_version} -ne 0 ] || [ ${include_reqs} -ne 0 ]
```
Comment (amousset) [inline tree/10_ncf_internals/list-compatible-inputs]: If we have `agent_version >= 3.9` and `agent_requirements  capabilities: [ dumpdatastate ]`, the file would not be included on 4.0. I think we need an && here, as we are matching negative conditions.

File: tree/10_ncf_internals/list-compatible-inputs (hunk 1 above)
Comment (peckpeck) [inline tree/10_ncf_internals/list-compatible-inputs]: yes we should

File: tree/10_ncf_internals/list-compatible-inputs (hunk 2 above)
Comment (peckpeck) [inline tree/10_ncf_internals/list-compatible-inputs]: Be careful, this is a negative expression, and it uses shell convention for true/false.

It reads: if we exclude for version or if we exclude for requirements, then exclude ...

**Comment by peckpeck** [comment]
Commit modified

**Comment by amousset** [comment]
I think we need to rename the cache file to add the capabilities.

**Comment by peckpeck** [comment]
Commit modified

**Pull request merged**